### PR TITLE
feat: post events to dataplane implementation

### PR DESCRIPTION
**Pull request by SKannaniOS**
## Description
This PR introduces the implementation for sending events to the data plane. This PR changes ensures the event ordering also.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Code refactor/optimization

## Implementation Details
This PR introduces the `FlushEvent` implementation, which triggers the posting of events to the data plane. Depending on the `Configuration` settings in `AnalyticsClient` object, events can be flushed either from the file system or directly from memory.

Note:
 - This PR provides a general Swift implementation related to the message handling, which will be adjusted in the future to meet the specific requirements for macOS, watchOS, and tvOS.

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added the necessary documentation (if appropriate).
- [x] I have ensured that my code follows the project's code style.
- [x] I have checked for potential performance impacts and optimized if necessary.
- [x] I have checked the code for security issues.
- [ ] I have updated the changelog (if required).

### CODE REVIEW

File: Analytics/Analytics/Helpers/Helpers.swift [hunk 1]
```diff
@@ -31,154 +32,14 @@ public struct Constants {
     static let batchPrefix = "{\"batch\":["
     static let batchSentAtSuffix = "],\"sentAt\":\""
     static let batchSuffix = "\"}"
-    static let configQueryParams = ["p": "ios", "v": "1.0.0"]
+    static let configQueryParams = ["p": "ios", "v": "1.29.1"]
```
Comment (1abhishekpandey) [inline Analytics/Analytics/Helpers/Helpers.swift:36]: Add a TODO to automate this hard-coded version update.

File: AnalyticsApp/AnalyticsApp/AnalyticsApp.swift [hunk 2]
```diff
@@ -39,7 +39,7 @@ class AnalyticsManager: Logger {
     private init() {}
     
     func initializeAnalyticsSDK() {
-        let config = Configuration(writeKey: "exmple_write_key", dataPlaneUrl: "https://data-plane.example.com", controlPlaneUrl: "https://control-plane.example.com", logger: self)
+        let config = Configuration(writeKey: "2h8ianoGGYaFRmDZ9s4Zx0n1GgD", dataPlaneUrl: "https://rudderstacfbtt.dataplane.rudderstack.com", logger: self)
```
Comment (1abhishekpandey) [inline AnalyticsApp/AnalyticsApp/AnalyticsApp.swift]: Please revert these changes.

File: Analytics/Analytics/Events/MessageManager.swift [hunk 3]
```diff
@@ -66,29 +69,95 @@ extension MessageManager {
...
+                self.storage.remove(messageReference: reference)
+            case .failure(let error):
+                print(error.localizedDescription)
+            }
+        })
+    }
+}
+
+extension MessageManager {
+    func isInOrder(_ batch: String) {
```
Comment (1abhishekpandey) [inline Analytics/Analytics/Events/MessageManager.swift]: What's the purpose of this `isInOrder` method?

Comment (1abhishekpandey) [reply Analytics/Analytics/Events/MessageManager.swift]: I guess all the methods defined under `extension MessageManager` will be removed!

If that's the case, please add a relevant comment, as a TODO.

File: Analytics/Analytics/Base/AnalyticsClient.swift [hunk 4]
```diff
@@ -50,22 +45,46 @@ extension AnalyticsClient {
...
+            case .failure(let error):
+                print(error.localizedDescription)
+            }
+        }
+    }
+}
+
+extension AnalyticsClient {
+    
+    var anonymousId: String {
```
Comment (1abhishekpandey) [inline Analytics/Analytics/Base/AnalyticsClient.swift:86]: I think we should not add the `anonymousId` for now, as it is part of state management. 

For now, we can use hard-coded strings with any random values and add a TODO. When we'll implement the state management, we can add the required anonymousId logic.

**Review by 1abhishekpandey — APPROVED** [review]

File: Analytics/Analytics/Helpers/Extensions.swift [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+//  Analytics
+//
+//  Created by Satheesh Kannan on 21/10/24.
+//
+
+import Foundation
+import zlib
+
+// MARK: - String
+extension String {
```
Comment (ChryssaAliferi) [inline Analytics/Analytics/Helpers/Extensions.swift:12]: I feel the `Extensions.swift` file is getting exponentially bigger. Is there a way to separate the contents in more than one Extension file or is this a widespread practice on swift that I am not aware of?

File: Analytics/Analytics/Base/AnalyticsClient.swift (hunk 4 above)
Comment (SKannaniOS) [reply Analytics/Analytics/Base/AnalyticsClient.swift:86]: This implementation is agreed for now & will be updated as per need in future.

File: Analytics/Analytics/Storage/Storage.swift [hunk 6]
```diff
@@ -46,19 +46,20 @@ public enum StorageMode: Int {
...
-    private static let shared = StorageQueue()
-    
-    private let queue = DispatchQueue(label: "rudderstack.storage.message.queue")
+    private static let semaphore = DispatchSemaphore(value: 1)
+    private static let queue = DispatchQueue(label: "rudderstack.message.storage.queue")
     
-    static func perform(_ action: @escaping StorageAction) {
-        Self.shared.queue.async(execute: action)
+    static func perform(_ action: @escaping () -> Void) {
+        self.queue.async {
```
Comment (1abhishekpandey) [inline Analytics/Analytics/Storage/Storage.swift:59]: The name of the class is `SynchronizedQueue` and over here we are using `async`. Do we need to use `sync`?

**Review by 1abhishekpandey — APPROVED** [review]

File: Analytics/Analytics/Storage/Storage.swift (hunk 6 above)
Comment (SKannaniOS) [reply Analytics/Analytics/Storage/Storage.swift:59]: I renamed the class to give it a more meaningful name.

**Review by 1abhishekpandey — APPROVED** [review]

**Review by ChryssaAliferi — APPROVED** [review]
I tried the branch out, seem great to me 🚀 
Let's proceed!

File: Analytics/Analytics/Helpers/Extensions.swift (hunk 5 above)
Comment (SKannaniOS) [reply Analytics/Analytics/Helpers/Extensions.swift:12]: Using a single extension file offers the advantage of managing all extensions in one place. Additionally, having separate files for each extension would increase the SDK size, so I prefer this approach.

**Review by ChryssaAliferi — APPROVED** [review]
